Fix/http request multiple params

Achyuthan-S opened the pull request:
## Summary

This PR improves the HTTP Request node's handling of multiple query parameters by fixing null value handling and adding proper binary data detection. The changes ensure consistent array handling while maintaining correct behavior for binary data objects.

Key improvements:
- Fixed null value misclassification
- Added strict binary data object validation
- Improved array parameter handling
- Enhanced code documentation

## Related Linear tickets, Github issues, and Community forum posts

Fixes #19727

## Review / Merge checklist

- [x] PR title and summary are descriptive
- [x] Tests included (tested with various parameter combinations including null values and binary data)
- [x] Docs updated (no documentation changes needed as this fixes existing behavior)
- [x] PR Labeled appropriately

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enhances HTTP Request V3 to correctly aggregate repeated params (incl. binary form-data), default auth to none, harden headers/response handling and UI sanitization, plus fixes array checks in fixedCollection processing.
> 
> - **HTTP Request V3 (`packages/nodes-base/nodes/HttpRequest/V3/HttpRequestV3.node.ts`)**
>   - Robust multi-parameter aggregation with binary form-data support; prevents overwrites and preserves multiple values.
>   - Default `authentication` to `none` when missing; import `Readable` at runtime.
>   - Safer header setup (`headers ??=`) and raw content-type assignment.
>   - Stronger typing for request/response objects; optional chaining for `headers['content-type']`.
>   - Sanitize-and-send UI messages with error logging instead of silent failures.
> - **Workflow helpers (`packages/workflow/src/node-helpers.ts`)**
>   - Validate `fixedCollection` `multipleValues` entries are arrays before iterating, avoiding invalid iterations.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 1db4d9293302bb2212ec6b820d4e31f5db68f122. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Achyuthan-S committed:
fix(HTTP Request): handle multiple values consistently in query parameters

Achyuthan-S committed:
fix(HTTP Request): improve handling of duplicate query parameters

Achyuthan-S committed:
fix(HTTP Request): improve binary data handling in query parameters

Achyuthan-S committed:
docs(HTTP Request): improve comment clarity for parameter handling

Achyuthan-S committed:
fix(HTTP Request): add strict guards for binary data object detection

Achyuthan-S committed:
docs(HTTP Request): clarify binary data validation in comments

Achyuthan-S committed:
docs: update comment to trigger fresh code review [trigger-review]

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=20277) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=20277) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=20277) it.</sub>

Achyuthan-S commented:
Hi n8n team! 

I've implemented the fix for issue #19727 by:

1. Adding strict guards for binary data object detection:
   - Added null check
   - Added proper object type validation
   - Added checks for required properties ('value' and 'options')

2. Improved the handling of duplicate parameters:
   - Fixed null value handling
   - Maintained array behavior for multiple values
   - Properly validated binary data objects before overwriting

The changes have been tested with various scenarios including null values, multiple parameters, and binary data objects.

Looking forward to your review and feedback!

Achyuthan-S committed:
fix(HTTP Request): Improve binary data handling in parameters

- Added proper type validation for binary data objects
- Fixed parameter accumulation to prevent binary data overwrite
- Preserved both binary and non-binary values when needed

Achyuthan-S commented on packages/nodes-base/nodes/HttpRequest/V3/HttpRequestV3.node.ts:
```diff
@@ -408,7 +408,29 @@ export class HttpRequestV3 implements INodeType {
+							'options' in existingValue
+						) {
+							// It's a validated binary data object (has value and options properties), overwrite it
+							accumulator[cur.name] = cur.value;
```
I have fixed the issue with binary data handling in the HTTP Request node. Here's what the new implementation does:

1. Added proper type validation for binary data objects by checking both `value` and `options` properties
2. Only overwrites existing binary data when the new value is also a validated binary object
3. If a non-binary parameter with the same name is encountered, it converts to an array to preserve both values
4. Added type guards to ensure proper handling of both binary and non-binary data

The new implementation ensures that binary payloads and metadata are never accidentally dropped, and both binary and non-binary values with the same parameter name are properly preserved.

For example:
- If we have a binary file parameter named "file" and later receive a non-binary parameter also named "file", both values will be preserved in an array
- Only when we receive another binary parameter named "file" will the previous binary value be overwritten

This should address the concern about preserving binary payloads while maintaining proper type safety.

Achyuthan-S committed:
fix(HttpRequest): Fix HTTP header names to use kebab-case format

- Fix content-length and content-type headers to use proper kebab-case format instead of camelCase
- Update IResponse interface to use correct 'content-type' header name
- Add null safety check for response.headers access
- Ensures HTTP standard compliance for binary data uploads and content type detection

Achyuthan-S committed:
fix: handle multiple binary parameters with same name in HTTP Request node

- Fixed bug where isNewBinary(cur.value) always evaluated to false
- Restructured parametersToKeyValue function to handle binary data properly
- Binary data preparation now happens before multiple values logic
- Fixed array handling to use spread operator instead of concat
- Ensures proper request body construction for form data with duplicate parameter names

Achyuthan-S committed:
fix: correct Bearer token authentication in HttpRequestV3

- Replace JSON.stringify() with String() for Bearer token conversion
- Ensures proper Authorization header format: 'Bearer <token>'
- Matches implementation in HttpRequestV2 for consistency
- Prevents malformed headers when token is unexpectedly an object

Achyuthan-S committed:
fix(HttpRequest): Make Readable stream detection robust across contexts; avoid instanceof and use constructor name + duck typing; keep Buffer instanceof check. Also minor safety/type improvements and headers init.

Achyuthan-S committed:
fix(HttpRequest): make isExistingBinary stream detection robust using duck-typing and Buffer.isBuffer

Achyuthan-S committed:
fix(HttpRequest): preserve multiple same-name binary form-data params as array instead of overwriting; default auth to 'none' if missing; minor typings and headers init fixes

Achyuthan-S committed:
Merge branch 'master' into fix/http-request-multiple-params

Achyuthan-S committed:
Merge branch 'master' into fix/http-request-multiple-params

Achyuthan-S committed:
Merge branch 'master' into fix/http-request-multiple-params
